client/src/api/block_builder/input_selection/new/mod.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub(crate) mod builder;
+pub(crate) mod burn;
+pub(crate) mod requirement;
+
+use std::collections::HashSet;
+
+use burn::Burn;
+use requirement::Requirements;
+
+use crate::{
+    block::{
+        address::{Address, AliasAddress, NftAddress},
+        output::{Output, OutputId},
+        protocol::ProtocolParameters,
+    },
+    error::{Error, Result},
+    secret::types::InputSigningData,
+};
+
+pub struct InputSelection {
+    outputs: Vec<Output>,
+    // TODO impl Iter ?
+    available_inputs: Vec<InputSigningData>,
+    protocol_parameters: ProtocolParameters,
+    timestamp: u32,
+    required_inputs: HashSet<OutputId>,
+    forbidden_inputs: HashSet<OutputId>,
+    remainder_address: Option<Address>,
+    burn: Option<Burn>,
+    // TODO: decide if we want to add the addresses here to check if we can unlock an output or not:
+    // alias output can have two different addresses and expiration unlock condition can change the unlock address
+    // sender_addresses: Vec<Address>,
+}
+
+impl InputSelection {
+    pub fn filter(mut self, addresses: &[Address]) -> Self {
+        let addresses = addresses
+            .iter()
+            // TODO meh
+            .copied()
+            .chain(self.available_inputs.iter().filter_map(|input| match &input.output {
+                Output::Alias(output) => Some(Address::Alias(AliasAddress::from(
+                    output.alias_id().or_from_output_id(*input.output_id()),
+                ))),
```

Comment:
I think here we need to over the outputs and check the state transition, if state index is the same as in the input, then it's a governor transition and then the address isn't unlocked
In all other cases the input selection would add them as state controller transition and then the address would be unlocked